bootstrap2.go

```diff
@@ -0,0 +1,100 @@
+package coopdatadog
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/coopnorge/go-datadog-lib/v2/internal"
+	datadogLogger "github.com/coopnorge/go-logger/adapter/datadog"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+	"gopkg.in/DataDog/dd-trace-go.v1/profiler"
+)
+
+// Start the Datadog integration, use the returned context.CancelFunc to stop
+// the Datadog integration.
+func Start(ctx context.Context, options ...Option) (context.Context, context.CancelFunc, error) {
```

Comment:
I understand this is so that the linter picks up the lack of cancellation / clean up?
It is a bit unclear to me how this differs from the following: 

```go
func main() {
  ctx, cancel := context.WithCancel(context.Background())
  // linter error, context is never cancelled.
  
  err := datadog.Start(ctx)
  if err != nil {
    panic(err)
  }
}

func Start(ctx context.Context) error {
  // init foo
  // init bar
  go func() {
    <-cxt.Done()
    tracer.Stop()
	profiler.Stop()
  }
  return err
}
```

Reply:
Th main difference is that `context.Context` does not guarantee a cancellable context.

The signature of this function is the  thing to figure out in this PR. We must somehow figure out to provide guarantees that proper shutdown is executed.

I'm open to suggestions :)

Reply:
In any case, we really ought to clearly document the following: 

- The function honors the context cancellation of the input context.
- What happens if the input context is cancelled already when calling the returned cancel function.
- The returned context get cancelled by the returned cancel function.

Reply:
Might this be more correct,

```suggestion
// Function signature for the shutdown function returned from Start
type DatadogStop func()

// var noopCancel = context.CancelFunc(func() {})
var noopCancel = func() {}

// Start the Datadog integration, use the returned context.CancelFunc to stop
// the Datadog integration.
func Start(_ context.Context, options ...Option) (DatadogStop, error) {
```

Go and linters would stil catch if we are not using the returned values.

Reply:
@hadrienk, @dalryan and I discussed the same topic as well and I landed on the same conclusion. I think a cancellation function is best as I do not know of a way to verify at compile time that a context can be canceled.

I will leave it to @dalryan to give his own conclusions.

Updates are coming.

Reply:
Rewritten to:

```go
// Start the Datadog integration, use the returned Cancel function to stop the
// Datadog integration. When calling the Cancel function traces will be flushed
// and profiling will be stopped to Datadog.
//
// Usage:
//
//	cancel, err := Start()
//	if err != nil {
//		panic(err)
//	}
//	defer func() {
//		err := cancel()
//		if err != nil {
//			panic(err)
//		}
//	}}
func Start(opts ...Option) (Cancel, error) {
	if internal.IsDatadogDisabled() {
		return noop, nil
	}

	err := internal.VerifyEnvVarsSet(
		internal.DatadogAPMEndpoint,
		internal.DatadogDSDEndpoint,
		internal.DatadogService,
		internal.DatadogEnvironment,
		internal.DatadogVersion,
	)
	if err != nil {
		return noop, err
	}

	options, err := resolveOptions(opts)
	if err != nil {
		return noop, err
	}

	l, err := datadogLogger.NewLogger(datadogLogger.WithGlobalLogger())
	if err != nil {
		return noop, fmt.Errorf("Failed to initialize the Datadog logger: %w", err)
	}
	ddtrace.UseLogger(l)

	cancel := func() error {
		return stop()
	}

	err = start(options)
	return cancel, err
}

// Cancel is a function signature for functions that stops the Datadog
// integration.
type Cancel func() error

func noop() error {
	return nil
}

var _ Cancel = noop
```

Should we allow for a custom cancel func that returns an error? This is useful since flushing metrics can error, see #434. Alternatively we handle stop errors using a error handler.

Reply:
The issue with the following block:

```go
  go func() {
    <-ctx.Done()
    cleanUp()
  }
```

inside the Start-function, is that it will start a cleanup at the same time as the application is shutting down, and there is no synchronization-point, so you won't be able to ensure that you have tried to flush the traces/profiling/metrics before the application exits.

A `Cancel` function (either `context.Cancel` or a custom `Cancel` function) would allow the caller of `Start` to ensure the proper shutdown of the datadog-clients. The same could be achieved with a separate function that did the same:

```go
// in main.go
func main(){
	if err := run(); err != nil {
		logger.WithError(err).Fatal("Unexpected application error") 
	}
	logger.Info("Done") 
}

func run() error{
	datadogOpts := ...
	err := datadog.Start(ctx, datadogOpts...)
	if err != nil {
		return err
	}
	defer datadog.StopAndFlush(ctx)
	
	// Rest of application
}
```

That way, it would be up to the caller to ensure that the StopAndFlush is called, but the result is the same. I still prefer a custom cancellation-function in the return values, instead of having to know from documentation which function to call to stop.

Reply:
I also think we should accept a `context.Context` in the `Start` function, to future proof ourselves (instead of having to add a option to pass a context if we need it in the future).
The current iteration of `dd-trace-go` does not require a context to start the tracer&friends, but it's not far-fetched to think that they might add that later on, to ensure that it can talk to the agent/otel-collector before marking itself as ready.

Reply:
Sounds good, I have added back the `ctx context.Context` parameter to `Start`. The signature is now,

```go
// Start the Datadog integration, use the returned Cancel function to stop the
// Datadog integration. When calling the Cancel function traces will be flushed
// and profiling will be stopped to Datadog.
//
// Usage:
//
//	package main
//
//	import (
//		"github.com/coopnorge/go-datadog-lib"
//	)
//
//	cancel, err := coopdatadog.Start(context.Background())
//	if err != nil {
//		panic(err)
//	}
//	defer func() {
//		err := cancel()
//		if err != nil {
//			panic(err)
//		}
//	}
func Start(ctx context.Context, opts ...Option) (Cancel, error) {
	...
}

// Cancel is a function signature for functions that stops the Datadog
// integration.
type Cancel func() error
```

Reply:
I think this is a sensible signature - and I agree with keeping the ctx in case we need it later on.

I also like the idea of returning the cancel function for guaranteeing the flush/cleanup. This seems to accomplish what we're aiming for pretty cleanly.

Reply:
Looks good! Maybe document clearly that the cancel function blocks?

Nitpicking here so feel free to ignore, maybe follow the `Func` suffix Go convention for function type?
Also, maybe StopFunc or ShutdownFunc would be more appropriate?

Reply:
`StopFunc` or `ShutdownFunc` would be inline with `context.CancelFunc` so I'm good with that. We currently do not have a guideline for typed function signatures.

@bendiknesbo, @hadrienk and @ryan what do we pick? `Stop`, `Shutdown`, `Cancel` with or without the suffix `Func`.

Reply:
`StopFunc` sounds good to me.

Reply:
This seems resolved.